### PR TITLE
Fix possible concurrent issue with SSL & default connector

### LINKED ISSUE #5797: Jersey core client 3.1.9 memory leak
If I'm not missing something this is a memory leak if you target unique URLs.

https://github.com/eclipse-ee4j/jersey/blob/f35691f4e4f1e6440ba2d384de798c4efdb2f932/core-client/src/main/java/org/glassfish/jersey/client/HttpUrlConnectorProvider.java#L301

### PR DESCRIPTION
**Pull request by jansupol**

### CODE REVIEW

File: tests/e2e-tls/src/test/java/org/glassfish/jersey/tests/e2e/tls/connector/ConcurrentHttpsUrlConnectionTest.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+ * all parallel requests from other Jersey client instances fail
+ * with SSLHandshakeException: PKIX path building failed.
+ * <p>
+ * Once the first GET request is completed,
+ * all subsequent requests work without error.
+ * <p>
+ * BUG 5749
+ */
+public class ConcurrentHttpsUrlConnectionTest {
+    private static int THREAD_NUMBER = 2;
```
Comment (dtbaum) [inline tests/e2e-tls/src/test/java/org/glassfish/jersey/tests/e2e/tls/connector/ConcurrentHttpsUrlConnectionTest.java]: From my perspective, only 2 threads feels a bit too minimal for a concurrency test.

File: core-client/src/main/java/org/glassfish/jersey/client/internal/HttpUrlConnector.java [hunk 2]
```diff
@@ -387,6 +387,10 @@ private ClientResponse _apply(final ClientRequest request) throws IOException {
             sniUri = request.getUri();
         }
 
+        if (!DEFAULT_SSL_SOCKET_FACTORY.isInitialized() && "HTTPS".equalsIgnoreCase(sniUri.getScheme())) {
```
Comment (dtbaum) [inline core-client/src/main/java/org/glassfish/jersey/client/internal/HttpUrlConnector.java:390]: The check !DEFAULT_SSL_SOCKET_FACTORY.isInitialized() seems to be not thread-save?

Comment (jansupol) [reply core-client/src/main/java/org/glassfish/jersey/client/internal/HttpUrlConnector.java:390]: The variable used in `isInitialized()` is volatile, so it should be thread-safe.

File: tests/e2e-tls/src/test/java/org/glassfish/jersey/tests/e2e/tls/connector/ConcurrentHttpsUrlConnectionTest.java (hunk 1 above)
Comment (jansupol) [reply tests/e2e-tls/src/test/java/org/glassfish/jersey/tests/e2e/tls/connector/ConcurrentHttpsUrlConnectionTest.java]: The premise is that the first request passes and all the consecutive do not until the first is finished. Two seem to comply with the premise and the test failed indeed without the fix. I can certainly put back your 5, if you insist.

**Review by senivam — APPROVED** [review]

**Review by jbescos — APPROVED** [review]

**Pull request merged**